cairo/ethereum/crypto/kzg.cairo

```diff
@@ -99,6 +121,184 @@ func bytes_to_bls_field{range_check_ptr, bitwise_ptr: BitwiseBuiltin*}(b: Bytes3
+    alloc_locals;
+
+    // Compute Q * (X - z)
+    tempvar n = U384(new UInt384(bls12_381.N0, bls12_381.N1, bls12_381.N2, bls12_381.N3));
+    let z_uint384 = uint256_to_uint384([z.value]);
+    tempvar z_u384 = U384(new z_uint384);
+    let neg_z = sub(n, z_u384, n);
+    let g2 = BLSP2_G();
+    let neg_z_g2 = blsp2_mul_by(g2, neg_z);
+    let signature_g2 = SIGNATURE_G2();
+    // x_minus_z values differ between EELS and Cairo implementation (debugging, to be confirmed).
+    let x_minus_z = blsp2_add(signature_g2, neg_z_g2);
+
+    // Compute P - y
+    let y_uint384 = uint256_to_uint384([y.value]);
+    tempvar y_u384 = U384(new y_uint384);
+    let neg_y = sub(n, y_u384, n);
+    let g1 = BLSP_G();
+    let neg_y_g1 = blsp_mul_by(g1, neg_y);
+    let (pubkey_from_commitment, error) = pubkey_to_g1(commitment);
+    assert error.value = 0;
+    let p_minus_y = blsp_add(pubkey_from_commitment, neg_y_g1);
+
+    // Compute -g2
+    let blsf2_zero = BLSF2_ZERO();
+    let neg_y_g2 = blsf2_sub(blsf2_zero, g2.value.y);
+    tempvar neg_g2 = BLSP2(new BLSP2Struct(g2.value.x, neg_y_g2));
+    // Compute pubkey_from_proof
+    let (pubkey_from_proof, error) = pubkey_to_g1(proof);
+    assert error.value = 0;
+
+
+    // Pairing check
+    let pairing_ptr: G1G2Pair* = alloc();
+    // 1st pairing: (p_minus_y, -g2)
+    // Convert to g1, g2 and G1G2Pair
+    let p_minus_y_garaga = G1PointGaraga(
+        [p_minus_y.value.x.value.c0.value], [p_minus_y.value.y.value.c0.value]
+    );
+    let neg_g2_garaga = G2Point(
+        [neg_g2.value.x.value.c0.value],
+        [neg_g2.value.x.value.c1.value],
+        [neg_g2.value.y.value.c0.value],
+        [neg_g2.value.y.value.c1.value],
+    );
+    assert [pairing_ptr] = G1G2Pair(p_minus_y_garaga, neg_g2_garaga);
+
+    // 2nd pairing: (pubkey_from_proof, x_minus_z)
+    let pubkey_proof_garaga = G1PointGaraga(
+        [pubkey_from_proof.value.x.value.c0.value], [pubkey_from_proof.value.y.value.c0.value]
+    );
+    let x_minus_z_garaga = G2Point(
+        [x_minus_z.value.x.value.c0.value],
+        [x_minus_z.value.x.value.c1.value],
+        [x_minus_z.value.y.value.c0.value],
+        [x_minus_z.value.y.value.c1.value],
+    );
+    assert [pairing_ptr + G1_G2_PAIR_SIZE] = G1G2Pair(pubkey_proof_garaga, x_minus_z_garaga);
+
+    let (pairing_check) = multi_pairing_2P(pairing_ptr);
+
+    // TODO:
+    // - Replace multi_pairing_2P by multi_miller_loop_1P + point at infinity + final exp
+    // - Why ? Ethereum KZG allows point at infinity, which is not handled in BLS standards, then not in Garaga
+    // - For each pairing
+    //     If one point of the pairing is point at infinity, return one (E12D_ONE)
+    //     Else apply miller loop on the pairing
+    // - Multiply both results of the pairing
+    // - Apply final exponentiation
+    // Pairing check
+    // In other words, reimplement `pairing` from EELS with a miller loop circuit from garaga.
```

Comment:
The pairing check cannot be treated as a black box anymore. Let's implement the pairing check in a new PR. Easier to isolate the logic and test.

Reply:
in fact it can be simpler: 
```
final_exponentiation = final_exponentiate(
        pairing(p_q_1[1], p_q_1[0], final_exponentiate=False)
        * pairing(p_q_2[1], p_q_2[0], final_exponentiate=False)
    )
```

is equivalent to 
``` 
final_exponentiate(pairing(p_q_1[1], p_q_1[0], final_exponentiate=False)) * final_exponentiate(pairing(p_q_2[1], p_q_2[0], final_exponentiate=False))
```
So it could be:
```
if q infinity point:
   temp_q = 1
else
   temp_q = multi_pairing_1P(q)
if p infinity point:
    temp_p = 1
else
   temp_p = multi_pairing_1P(p)
assert  temp_p * temp_q == 1
```